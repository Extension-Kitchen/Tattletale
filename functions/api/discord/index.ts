/**
 * The core server that runs on a Cloudflare worker.
 */

import {
  InteractionResponseType,
  InteractionType,
  verifyKey,
} from 'discord-interactions';
import { AWW_COMMAND, INVITE_COMMAND } from '../../../src/discord/commands';
import { InteractionResponseFlags } from 'discord-interactions';
import { Env } from '../../../src/env';

class JsonResponse extends Response {
  constructor(body: unknown, init?: ResponseInit) {
    const jsonBody = JSON.stringify(body);
    init = init || {
      headers: {
        'content-type': 'application/json;charset=UTF-8',
      },
    };
    super(jsonBody, init);
  }
}

/**
 * Main route for all requests sent from Discord.  All incoming messages will
 * include a JSON payload described here:
 * https://discord.com/developers/docs/interactions/receiving-and-responding#interaction-object
 */
export const onRequestPost: PagesFunction<Env> = async (context) => {
  const { env, request } = context;
  const { isValid, interaction } = await verifyDiscordRequest(context);

  if (!isValid || !interaction) {
    return new Response('Bad request signature.', { status: 401 });
  }

  if (interaction.type === InteractionType.PING) {
    // The `PING` message is used during the initial webhook handshake, and is
    // required to configure the webhook in the developer portal.
    return new JsonResponse({
      type: InteractionResponseType.PONG,
    });
  }

  if (interaction.type === InteractionType.APPLICATION_COMMAND) {
    // Most user commands will come as `APPLICATION_COMMAND`.
    switch (interaction.data.name.toLowerCase()) {
      case AWW_COMMAND.name.toLowerCase(): {
        return new JsonResponse({
          type: InteractionResponseType.CHANNEL_MESSAGE_WITH_SOURCE,
          data: {
            content: "Something",
          },
        });
      }
      case INVITE_COMMAND.name.toLowerCase(): {
        const applicationId = env.DISCORD_APPLICATION_ID;
        const INVITE_URL = `https://discord.com/oauth2/authorize?client_id=${applicationId}&scope=applications.commands`;
        return new JsonResponse({
          type: InteractionResponseType.CHANNEL_MESSAGE_WITH_SOURCE,
          data: {
            content: INVITE_URL,
            flags: InteractionResponseFlags.EPHEMERAL,
          },
        });
      }
      default:
        return new JsonResponse({ error: 'Unknown Type' }, { status: 400 });
    }
  }

  console.error('Unknown Type');
  return new JsonResponse({ error: 'Unknown Type' }, { status: 400 });
};

async function verifyDiscordRequest({ request, env }: EventContext<Env, any, Record<string, unknown>>) {
  const signature = request.headers.get('x-signature-ed25519');
  const timestamp = request.headers.get('x-signature-timestamp');
  const body = await request.text();
  const isValidRequest =
    signature &&
    timestamp &&
    verifyKey(body, signature, timestamp, env.DISCORD_PUBLIC_KEY);
  if (!isValidRequest) {
    return { isValid: false };
  }

  return { interaction: JSON.parse(body), isValid: true };
}
