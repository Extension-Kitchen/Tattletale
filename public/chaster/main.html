<html>
	<script type="module">
		const inviteUri = await fetch("/api/discord/invite-uri").then(r => r.text());
		
		const anchor = document.querySelector("#discord-invite-link");
		anchor.href = inviteUri;
		anchor.innerHTML = inviteUri;
	</script>
	<body>
		<h1 style="color: #fff;">Discord bot invite link</h1>
		<a id="discord-invite-link" target="_blank" rel="noopener noreferrer" style="color: #fff;"></a>
	</body>
</html>