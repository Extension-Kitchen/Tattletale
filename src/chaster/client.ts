import createClient, { Middleware } from "openapi-fetch";
import type { paths } from "./api";
import { Env } from "@src/env";

export const baseApiUrl = "https://api.chaster.app";

let client = null;

export function useClient(env: Env): ReturnType<typeof createClient<paths>> {
  if (client === null) {
    const auth: Middleware = {
      async onRequest(req) {
        req.headers.set("Authorization", `Bearer ${env.CHASTER_TOKEN}`);
        return req;
      },
    };

    client = createClient<paths>({ baseUrl: baseApiUrl });
    client.use(auth);
  }

  return client;
}
