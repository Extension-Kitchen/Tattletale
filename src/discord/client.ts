import createClient from "openapi-fetch";
import type { paths } from "./api";

export const baseApiHost = "https://discord.com";
export const baseApiPath = "/api/v10";
export const baseApiUrl = `${baseApiHost}${baseApiPath}`;

export const client = createClient<paths>({ baseUrl: baseApiUrl });
