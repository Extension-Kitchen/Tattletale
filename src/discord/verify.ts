import { verifyKey } from "discord-interactions";
import { Env } from "../env";

// type VerifiedInteraction =
//   | {
//       interaction:
//         | components["schemas"]["PongInteractionCallbackRequest"]
//         | components["schemas"]["ApplicationCommandSubcommandGroupOption"]
//         | components["schemas"]["MessageComponentStringSelectResponse"];

//       isValid: true;
//     }
//   | {
//       interaction: undefined;
//       isValid: false;
//     };

export async function verifyDiscordRequest({
  request,
  env,
}: EventContext<Env, string, Record<string, unknown>>) {
  const signature = request.headers.get("x-signature-ed25519");
  const timestamp = request.headers.get("x-signature-timestamp");
  const body = await request.text();
  const isValidRequest =
    signature &&
    timestamp &&
    verifyKey(body, signature, timestamp, env.DISCORD_PUBLIC_KEY);
  if (!isValidRequest) {
    return { isValid: false, interaction: undefined };
  }

  return { interaction: JSON.parse(body), isValid: true };
}
