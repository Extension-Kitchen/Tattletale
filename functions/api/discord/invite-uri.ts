import { Env } from "@src/env";
import { inviteUri } from "@src/discord/invite-uri";

export const onRequestGet: PagesFunction<Env> = async (context) => {
  return new Response(inviteUri(context.env));
};
