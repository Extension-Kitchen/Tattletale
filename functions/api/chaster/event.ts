import { sendDm } from "@src/discord/messaging";
import { components } from "@src/chaster/api";
import { Env } from "@src/env";
import { useClient } from "@src/chaster/client";

type WebhookPayload = components["schemas"]["WebhookEventForPublic"]["payload"];
type WebhookPayloadDataMap = {
  [Property in WebhookPayload["event"]]: Extract<
    WebhookPayload,
    { event: Property }
  >["data"];
};

async function lockEvent(
  env: Env,
  lock: components["schemas"]["LockForPublic"],
  message: string,
) {
  const userDiscordSnowflake = lock.user.discordId;
  const khDiscordSnowflake = lock.keyholder.discordId;

  await Promise.all([
    sendDm(env, khDiscordSnowflake, message),
    sendDm(env, userDiscordSnowflake, message),
  ]);
}

async function actionLogEvent(
  env: Env,
  data: WebhookPayloadDataMap["action_log.created"],
) {
  const client = useClient(env);

  const session = await client.GET("/api/extensions/sessions/{sessionId}", {
    params: {
      path: {
        sessionId: data.sessionId,
      },
    },
  });

  await lockEvent(env, session.data?.session.lock, JSON.stringify(data));
}

async function sessionCreatedEvent(
  env: Env,
  data: WebhookPayloadDataMap["extension_session.created"],
) {
  await lockEvent(env, data.session.lock, JSON.stringify(data));
}

async function sessionUpdatedEvent(
  env: Env,
  data: WebhookPayloadDataMap["extension_session.updated"],
) {
  await lockEvent(env, data.session.lock, JSON.stringify(data));
}

async function sessionDeletedEvent(
  env: Env,
  data: WebhookPayloadDataMap["extension_session.deleted"],
) {
  await lockEvent(env, data.session.lock, JSON.stringify(data));
}

export const onRequestPost: PagesFunction<Env> = async (context) => {
  const data: WebhookPayload = await context.request.json();

  switch (data.event) {
    case "action_log.created":
      await actionLogEvent(context.env, data.data);
      break;
    case "extension_session.created":
      await sessionCreatedEvent(context.env, data.data);
      break;
    case "extension_session.updated":
      await sessionUpdatedEvent(context.env, data.data);
      break;
    case "extension_session.deleted":
      await sessionDeletedEvent(context.env, data.data);
      break;
    default:
      console.error(
        `unknown chaster event type ${(data as { event: string }).event}`,
      );
  }

  return new Response();
};
