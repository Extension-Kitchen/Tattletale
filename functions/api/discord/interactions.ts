/**
 * The discord interactions endpoint
 */

import { InteractionResponseType, InteractionType } from "discord-interactions";
import { INVITE_COMMAND } from "@src/discord/commands";
import { InteractionResponseFlags } from "discord-interactions";
import { Env } from "@src/env";
import { verifyDiscordRequest } from "@src/discord/verify";
import { inviteUri } from "@src/discord/invite-uri";

class JsonResponse extends Response {
  constructor(body: unknown, init?: ResponseInit) {
    const jsonBody = JSON.stringify(body);
    init = init || {
      headers: {
        "content-type": "application/json;charset=UTF-8",
      },
    };
    super(jsonBody, init);
  }
}

/**
 * Main route for all requests sent from Discord.  All incoming messages will
 * include a JSON payload described here:
 * https://discord.com/developers/docs/interactions/receiving-and-responding#interaction-object
 */
export const onRequestPost: PagesFunction<Env> = async (context) => {
  const { env } = context;
  const { isValid, interaction } = await verifyDiscordRequest(context);

  if (!isValid || !interaction) {
    return new Response("Bad request signature.", { status: 401 });
  }

  if (interaction.type === InteractionType.PING) {
    // The `PING` message is used during the initial webhook handshake, and is
    // required to configure the webhook in the developer portal.
    return new JsonResponse({
      type: InteractionResponseType.PONG,
    });
  }

  if (interaction.type === InteractionType.APPLICATION_COMMAND) {
    // Most user commands will come as `APPLICATION_COMMAND`.
    switch (interaction.data.name.toLowerCase()) {
      case INVITE_COMMAND.name.toLowerCase(): {
        return new JsonResponse({
          type: InteractionResponseType.CHANNEL_MESSAGE_WITH_SOURCE,
          data: {
            content: inviteUri(env),
            flags: InteractionResponseFlags.EPHEMERAL,
          },
        });
      }
      default:
        return new JsonResponse({ error: "Unknown Type" }, { status: 400 });
    }
  }

  console.error("Unknown Type");
  return new JsonResponse({ error: "Unknown Type" }, { status: 400 });
};
