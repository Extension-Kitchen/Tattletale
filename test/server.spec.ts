// NOTE: This link is a good example of pages functions and testing
// https://github.com/cloudflare/workers-sdk/tree/main/fixtures/vitest-pool-workers-examples/pages-functions-unit-integration-self

import {
  InteractionResponseType,
  InteractionType,
  InteractionResponseFlags,
} from "discord-interactions";
import { INVITE_COMMAND } from "@src/discord/commands";
import { SELF, fetchMock } from "cloudflare:test";
import {
  describe,
  it,
  expect,
  vi,
  beforeAll,
  afterEach,
  beforeEach,
} from "vitest";
import {
  baseApiHost as discordBaseApiHost,
  baseApiPath as discordBaseApiPath,
} from "@src/discord/client";
import { baseApiUrl as chasterBaseApiHost } from "@src/chaster/client";
import * as discordVerifyMod from "@src/discord/verify";

// To make hot reloading work
import * as _ from "../dist-functions";

vi.mock("@src/discord/verify");

beforeAll(() => {
  // Enable outbound request mocking...
  fetchMock.activate();
  // ...and throw errors if an outbound request isn't mocked
  fetchMock.disableNetConnect();
});

beforeEach(() => {
  vi.resetModules();
  vi.resetAllMocks();
});

// Ensure we matched every mock we defined
afterEach(() => fetchMock.assertNoPendingInterceptors());

describe("Server", () => {
  describe("GET /", () => {
    it("should return a greeting message with the Discord application ID", async () => {
      const response = await SELF.fetch("https://example.com/api");
      const body = await response.text();

      expect(body).to.satisfy((b) => b.startsWith("ðŸ‘‹"));
    });
  });

  describe("POST /", () => {
    it("should handle a PING interaction", async () => {
      const interaction = {
        type: InteractionType.PING,
      };

      vi.mocked(discordVerifyMod.verifyDiscordRequest).mockReturnValue(
        Promise.resolve({
          isValid: true,
          interaction: interaction,
        }),
      );

      const response = await SELF.fetch(
        "https://example.com/api/discord/interactions",
        {
          method: "POST",
        },
      );
      const body = await response.json();
      expect(body["type"]).to.equal(InteractionResponseType.PONG);
    });

    it("should handle a lock event", async () => {
      const channelId = 0;
      const snowflake = 1243;
      const sessionId = 143;

      // mock the fetch calls to chaster
      fetchMock
        .get(chasterBaseApiHost)
        .intercept({
          method: "GET",
          headers: (headers) => headers["authorization"].startsWith("Bearer "),
          path: `/api/extensions/sessions/${sessionId}`,
        })
        .reply(200, {
          session: {
            lock: {
              user: { discordId: snowflake },
              keyholder: { discordId: snowflake },
            },
          },
        });

      // mock the fetch calls to discord
      for (const _ of [0, 1]) {
        fetchMock
          .get(discordBaseApiHost)
          .intercept({
            method: "POST",
            body: (b) => JSON.parse(b)["recipient_id"] === snowflake,
            headers: (headers) => headers["authorization"].startsWith("Bot "),
            path: discordBaseApiPath + "/users/@me/channels",
          })
          .reply(200, { id: channelId });

        fetchMock
          .get(discordBaseApiHost)
          .intercept({
            method: "POST",
            headers: (headers) => headers["authorization"].startsWith("Bot "),
            path: discordBaseApiPath + `/channels/${channelId}/messages`,
          })
          .reply(200, {});
      }

      const response = await SELF.fetch(
        "https://example.com/api/chaster/event",
        {
          method: "POST",
          body: JSON.stringify({
            event: "action_log.created",
            data: { sessionId: sessionId },
          }),
        },
      );

      expect(response.status).to.equal(200);
    });

    it("should handle an invite command interaction", async () => {
      const interaction = {
        type: InteractionType.APPLICATION_COMMAND,
        data: {
          name: INVITE_COMMAND.name,
        },
      };

      vi.mocked(discordVerifyMod.verifyDiscordRequest).mockReturnValue(
        Promise.resolve({
          isValid: true,
          interaction: interaction,
        }),
      );

      const response = await SELF.fetch(
        "https://example.com/api/discord/interactions",
        {
          method: "POST",
        },
      );
      const body = await response.json();
      expect(body["type"]).to.equal(
        InteractionResponseType.CHANNEL_MESSAGE_WITH_SOURCE,
      );
      expect(body["data"].content).to.matchSnapshot();
      expect(body["data"].flags).to.equal(InteractionResponseFlags.EPHEMERAL);
    });

    it("should handle an unknown command interaction", async () => {
      const interaction = {
        type: InteractionType.APPLICATION_COMMAND,
        data: {
          name: "unknown",
        },
      };

      vi.mocked(discordVerifyMod.verifyDiscordRequest).mockReturnValue(
        Promise.resolve({
          isValid: true,
          interaction: interaction,
        }),
      );

      const response = await SELF.fetch(
        "https://example.com/api/discord/interactions",
        {
          method: "POST",
        },
      );
      expect(
        vi.mocked(discordVerifyMod.verifyDiscordRequest).mock.calls.length,
      ).to.equal(1);
      const body = await response.json();
      expect(response.status).to.equal(400);
      expect(body["error"]).to.equal("Unknown Type");
    });
  });

  describe("All other routes", () => {
    it('api should return a "Not Found" response', async () => {
      const response = await SELF.fetch("https://example.com/api/unknown");
      expect(response.status).to.equal(404);
      const body = await response.text();
      expect(body).to.equal("");
    });

    it('pages should return a "Not Found" response', async () => {
      const response = await SELF.fetch("https://example.com/unknown");
      expect(response.status).to.equal(404);
      const body = await response.text();
      expect(body).to.equal("");
    });
  });
});
