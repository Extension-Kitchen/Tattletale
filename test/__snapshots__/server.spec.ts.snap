// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Server > POST / > should handle an invite command interaction 1`] = `"https://discord.com/oauth2/authorize?client_id=1250200555113025737"`;
